src/hotspot/share/gc/g1/g1CardSet.cpp

```diff
@@ -0,0 +1,899 @@
+
+  assert(is_power_of_2(_max_cards_in_card_set),
+        "max_cards_in_card_set must be a power of 2: %u", _max_cards_in_card_set);
+
+  _num_buckets_in_howl = G1CardSetHowl::num_buckets(_max_cards_in_card_set, _num_cards_in_array, max_buckets_in_howl);
+
+  _num_cards_in_howl_bitmap = G1CardSetHowl::bitmap_size(_max_cards_in_card_set, _num_buckets_in_howl);
+  _cards_in_howl_bitmap_threshold = _num_cards_in_howl_bitmap * cards_in_bitmap_threshold;
+  _log2_num_cards_in_howl_bitmap = log2i_exact(_num_cards_in_howl_bitmap);
+  _bitmap_hash_mask = ~(~(0) << _log2_num_cards_in_howl_bitmap);
+
+  log_configuration();
+}
+
+void G1CardSetConfiguration::log_configuration() {
+  log_debug_p(gc, remset)("Card Set container configuration: "
+                          "InlinePtr #elems %u size %zu "
+                          "Array Of Cards #elems %u size %zu "
+                          "Howl #buckets %u coarsen threshold %u "
+                          "Howl Bitmap #elems %u size %zu coarsen threshold %u",
+                          num_cards_in_inline_ptr(), sizeof(void*),
+                          num_cards_in_array(), G1CardSetArray::size_in_bytes(num_cards_in_array()),
+                          num_buckets_in_howl(), cards_in_howl_threshold(),
+                          num_cards_in_howl_bitmap(), G1CardSetBitMap::size_in_bytes(num_cards_in_howl_bitmap()), cards_in_howl_bitmap_threshold());
+}
+
+uint G1CardSetConfiguration::num_cards_in_inline_ptr() const {
+  return num_cards_in_inline_ptr(_inline_ptr_bits_per_card);
+}
+
+uint G1CardSetConfiguration::num_cards_in_inline_ptr(uint bits_per_card) {
+  return G1CardSetInlinePtr::max_cards_in_inline_ptr(bits_per_card);
+}
+
+G1CardSetAllocOptions* G1CardSetConfiguration::mem_object_alloc_options() {
+  G1CardSetAllocOptions* result = NEW_C_HEAP_ARRAY(G1CardSetAllocOptions, num_mem_object_types(), mtGC);
+
+  result[0] = { (uint)CardSetHash::get_node_size() };
+  result[1] = { (uint)G1CardSetArray::size_in_bytes(num_cards_in_array()), 2, 256 };
+  result[2] = { (uint)G1CardSetBitMap::size_in_bytes(num_cards_in_howl_bitmap()), 2, 256 };
+  result[3] = { (uint)G1CardSetHowl::size_in_bytes(num_buckets_in_howl()), 2, 256 };
+
+  return result;
+}
+
+const char* G1CardSetConfiguration::mem_object_type_name_str(uint index) {
+  const char* names[] = { "Node", "Array", "Bitmap", "Howl" };
+  return names[index];
+}
+
+void G1CardSetCoarsenStats::reset() {
+  STATIC_ASSERT(ARRAY_SIZE(_coarsen_from) == ARRAY_SIZE(_coarsen_collision));
+  for (uint i = 0; i < ARRAY_SIZE(_coarsen_from); i++) {
+    _coarsen_from[i] = 0;
+    _coarsen_collision[i] = 0;
+  }
+}
+
+void G1CardSetCoarsenStats::add(G1CardSetCoarsenStats& other) {
+  STATIC_ASSERT(ARRAY_SIZE(_coarsen_from) == ARRAY_SIZE(_coarsen_collision));
+  for (uint i = 0; i < ARRAY_SIZE(_coarsen_from); i++) {
+    _coarsen_from[i] += other._coarsen_from[i];
+    _coarsen_collision[i] += other._coarsen_collision[i];
+  }
+}
+
```

Comment:
Unused from what I can tell.
```suggestion
```

Reply:
Yes, after the change to add `subtract_from`...